src/txmempool.h

```diff
@@ -816,6 +790,111 @@ class CTxMemPool
         assert(m_epoch.guarded()); // verify guard even when it==nullopt
         return !it || visited(*it);
     }
+
+    /*
+     * CTxMemPoolChangeSet:
+     *
+     * This class is used for all mempool additions and associated removals (eg
+     * due to rbf). Removals that don't need to be evaluated for acceptance,
+     * such as removing transactions that appear in a block, or due to reorg,
+     * or removals related to mempool limiting or expiry do not need to use
+     * this.
+     *
+     * Callers can invoke StageAddition()/StageRemoval() of transactions
+     * in any order.
+     *
+     * CalculateChunksForRBF() can be used to calculate the feerate diagram of
+     * the proposed set of new transactions and compare with the existing
+     * mempool.
+     *
+     * CalculateMemPoolAncestors() calculates the in-mempool (not including
+     * what is in the change set itself) ancestors of a given transacion.
+     *
+     * Apply() will apply the removals and additions that are staged into the
+     * mempool.
+     *
+     * Only one changeset may exist at a time. While a changeset is
+     * outstanding, no removals or additions may be made directly to the
+     * mempool.
+     */
+    class CTxMemPoolChangeSet {
+    public:
+        explicit CTxMemPoolChangeSet(CTxMemPool* pool) : m_pool(pool) {}
+        ~CTxMemPoolChangeSet() EXCLUSIVE_LOCKS_REQUIRED(m_pool->cs) { m_pool->m_changeset = nullptr; }
+
+        CTxMemPoolChangeSet(const CTxMemPoolChangeSet&) = delete;
+        CTxMemPoolChangeSet& operator=(const CTxMemPoolChangeSet&) = delete;
+
+        using TxHandle = CTxMemPool::txiter;
+
+        TxHandle StageAddition(const CTransactionRef& tx, const CAmount fee, int64_t time, unsigned int entry_height, uint64_t entry_sequence, bool spends_coinbase, int64_t sigops_cost, LockPoints lp);
+        void StageRemoval(CTxMemPool::txiter it) { m_to_remove.insert(it); }
+
+        const CTxMemPool::setEntries& GetRemovals() const { return m_to_remove; }
+
+        util::Result<CTxMemPool::setEntries> CalculateMemPoolAncestors(TxHandle tx, const Limits& limits) {
```

Comment:
I reworked your suggestion to capture the caching idea, and now invoke this function in `Apply`.  However, there's still an ugly special case in `Apply()` in that we can only used a cached ancestor set calculation for the first transaction in a package (because the calculation is no longer correct once transactions start getting added to the mempool).

Note that in the cluster mempool implementation, this ugliness in `Apply()` goes away, because we will no longer need to have ancestor sets calculated in order to add a transaction to the mempool.  

Furthermore, I believe that we may be able to completely eliminate the need for this function altogether, post-cluster-mempool.  I should go back and try to rework #28676 to demonstrate that this is the case, but my recollection is that the only reason we will need to calculate ancestors sets in the future is for an RBF sanity check (namely that we're not replacing an ancestor of any to-be-added transaction) which we can reimplement without these ancestor calculations.

Reply:
Looking into whether I can avoid exposing this function even before cluster mempool lands; will consider it for a followup PR.